### PR TITLE
Added functionality to check if a voucher is available for basket

### PR DESCRIPTION
**Pull request by Ayub-Khan**
[Enhancement-2843](https://github.com/django-oscar/django-oscar/issues/2843)

Added functionality to check if a voucher is available for a basket or not.

Tests : Done
Docs : Not sure!

### CODE REVIEW

**Comment by solarissmoke** [comment]
I am not convince by the utility of this method. It seems like you can achieve the same thing just by calling `voucher.is_available_to_user(basket.owner)`?

**Comment by Ayub-Khan** [comment]
@solarissmoke  this can be used to replace the following code. https://github.com/django-oscar/django-oscar/blob/master/src/oscar/apps/basket/views.py#L340-L360 . I assume that if a voucher has valid offer it gives discount. Plus this can be used by other projects built on django-oscar and have to check voucher validation for basket in different views. This can also help them to override to add basket base checks. Also adding voucher and thn checking the discount and removing if not valid just to validate the voucher does not seem very appropriate.

File: src/oscar/apps/voucher/abstract_models.py [hunk 1]
```diff
@@ -219,6 +219,24 @@ def is_available_to_user(self, user=None):
                                 "a previous order")
         return is_available, message
 
+    def is_available_for_basket(self, basket):
+        """
+        Tests whether this voucher is available to the pased basket.
```
Comment (craigloftus) [inline src/oscar/apps/voucher/abstract_models.py]: You have a typo here, `pased`

File: src/oscar/apps/voucher/abstract_models.py [hunk 2]
```diff
@@ -219,6 +219,24 @@ def is_available_to_user(self, user=None):
...
+
+        Returns a tuple of a boolean for whether it is successful, and a
+        availability message.
+        """
+        is_available, message = False, _("This voucher is not available for Basket")
```
Comment (craigloftus) [inline src/oscar/apps/voucher/abstract_models.py]: Can you change this message to "This voucher is not available for this basket".

File: src/oscar/apps/voucher/abstract_models.py [hunk 3]
```diff
@@ -219,6 +219,24 @@ def is_available_to_user(self, user=None):
...
+        valid_user, __ = self.is_available_to_user(user=basket.owner)
+        if not valid_user:
+            message = _("This voucher is not available to user")
```
Comment (craigloftus) [inline src/oscar/apps/voucher/abstract_models.py]: And this should be using the message returned by `is_available_to_user`.

**Comment by Ayub-Khan** [comment]
@craigloftus Thanks for review. I have updated the pr. i did run the 'make docs' command. Are there any files where i will have to write those doc details or doc strings are enough.

**Comment by Ayub-Khan** [comment]
@specialunderwear can you review this too?. Let me know if there is any thing else i need to do.

**Comment by highbiza** [comment]
looks good to me

**Comment by specialunderwear** [comment]
looks good to me

**Comment by Ayub-Khan** [comment]
@specialunderwear what do i need to do to get this merged?

**Comment by specialunderwear** [comment]
Ask somebody that can actually commit to master?

**Pull request merged**